src/streamlink/plugins/pluto.py

```diff
@@ -1,99 +1,166 @@
+                },
+            ),
         )
 
-    def _get_streams(self):
-        data = None
+    def _get_playlist(self, host, path, params, token):
+        qs = parse_qs(params)
+        qs["jwt"] = token
+        yield from PlutoHLSStream.parse_variant_playlist(self.session, update_qsd(urljoin(host, path), qs)).items()
 
+    @staticmethod
+    def _get_media_data(data, key, slug):
+        media = data.get(key)
+        if media and media[0]["slug"] == slug:
+            return media[0]
+
+    def _get_streams(self):
         m = self.match.groupdict()
-        if m['slug_live']:
-            res = self.session.http.get('https://api.pluto.tv/v2/channels')
-            data = self.session.http.json(res,
-                                          schema=self._schema_media(m['slug_live']))
-        elif m['slug_series'] and m['slug_episode']:
-            res = self.session.http.get(f'http://api.pluto.tv/v3/vod/slugs/{m["slug_series"]}')
-            data = self.session.http.json(
-                res,
-                schema=validate.Schema(
-                    {'seasons': validate.all(
-                        [{'episodes': self._schema_media(m['slug_episode'])}],
-                        validate.filter(lambda k: k['episodes'] is not None))},
-                    validate.get('seasons'),
-                    validate.get(0),
-                    validate.any(None, validate.get('episodes'))
-                ),
-            )
-        elif m['slug_movies']:
-            res = self.session.http.get('https://api.pluto.tv/v3/vod/categories',
-                                        params={'includeItems': 'true', 'deviceType': 'web'})
-            data = self.session.http.json(
-                res,
-                schema=validate.Schema(
-                    {'categories': validate.all(
-                        [{'items': self._schema_media(m['slug_movies'])}],
-                        validate.filter(lambda k: k['items'] is not None))},
-                    validate.get('categories'),
-                    validate.get(0),
-                    validate.any(None, validate.get('items')),
-                ),
-            )
-
-        log.trace(f'{data!r}')
-        if data is None or not data.get('stitched'):
+        if m["slug_live"]:
+            data = self._get_api_data("channelSlug", m["slug_live"])
+            media = self._get_media_data(data, "EPG", m["slug_live"])
+            if not media:
+                return
+
+            self.id = media["id"]
+            self.title = media["name"]
+            path = media["stitched"]["path"]
+
+        elif m["slug_series"] and m["slug_episode"]:
+            data = self._get_api_data("episodeSlugs", m["slug_series"], filter=m["slug_episode"])
+            media = self._get_media_data(data, "VOD", m["slug_series"])
+            if not media or "seasons" not in media:
+                return
+            episode = media["seasons"][0]["episodes"][0]
```

Comment:
`media["seasons"][0]["episodes"][0]` doesn't work if you try to watch a show that's not part of season 1 (index 0). Every `episodes` list will be empty other than the one of the current seasons object if it's not the first season.

```
$ streamlink -l debug 'https://pluto.tv/en/on-demand/series/south-park-de/season/13/episode/der-ring-2008-13-1301' best
[cli][debug] OS:         Linux-5.15.5-2-git-x86_64-with-glibc2.33
[cli][debug] Python:     3.9.9
[cli][debug] Streamlink: 3.0.3+3.g59e8669
[cli][debug] Requests(2.26.0), Socks(1.7.1), Websocket(1.2.1)
[cli][debug] Arguments:
[cli][debug]  url=https://pluto.tv/en/on-demand/series/south-park-de/season/13/episode/der-ring-2008-13-1301
[cli][debug]  stream=['best']
[cli][debug]  --loglevel=debug
[cli][debug]  --player=mpv
[cli][info] Found matching plugin pluto for URL https://pluto.tv/en/on-demand/series/south-park-de/season/13/episode/der-ring-2008-13-1301
[plugins.pluto][debug] slug=south-park-de
[plugins.pluto][debug] app_version=5.106.0-f3e2ac48d1dbe8189dc784777108b725b4be6be2
Traceback (most recent call last):
  File "/home/basti/.local/bin/streamlink", line 33, in <module>
    sys.exit(load_entry_point('streamlink', 'console_scripts', 'streamlink')())
  File "/home/basti/repos/streamlink/src/streamlink_cli/main.py", line 1056, in main
    handle_url()
  File "/home/basti/repos/streamlink/src/streamlink_cli/main.py", line 565, in handle_url
    streams = fetch_streams(plugin)
  File "/home/basti/repos/streamlink/src/streamlink_cli/main.py", line 459, in fetch_streams
    return plugin.streams(stream_types=args.stream_types,
  File "/home/basti/repos/streamlink/src/streamlink/plugin/plugin.py", line 336, in streams
    ostreams = self._get_streams()
  File "/home/basti/repos/streamlink/src/streamlink/plugins/pluto.py", line 131, in _get_streams
    episode = media["seasons"][0]["episodes"][0]
IndexError: list index out of range
```

Reply:
Ah, well spotted, thanks.  I must have checked this only in cases where the season happened to be at index 0, which of course is unrelated to the actual season number, which is what I was going by.

Is there any way to filter the empty episodes within the schema also?  I'm having trouble trying to work out how that might be achieved, if it's possible.  Perhaps it might be better to loop through the episodes checking the slug of each for a match within the main plugin code instead?

Reply:
I'd say that the validation schema should be kept as simple as possible.